### ISSUE TITLE
passportConfigurator userModel not responding to change

### ISSUE THREAD

**Issue by ianseyer** [strongloop/loopback-component-passport#95]
Hi there!

I am using the facebook passport component, to handle user signup via facebook.

It is fully functional, however:

despite my server.js containing:

```
passportConfigurator.setupModels({
 userModel: app.models.donor,
 userIdentityModel: app.models.userIdentity,
 userCredentialModel: app.models.userCredential
});
```
the users are still created in the standard User table, instead of my Donor table.

**Comment by superkhau** [comment]
Duplicate. See https://github.com/strongloop/loopback/issues/1750

**Comment by superkhau** [comment]
Keeping this issue as the main and closing https://github.com/strongloop/loopback/issues/1750 as a duplicate.

**Comment by clockworkgr** [comment]
Following the discussion you had at strongloop/loopback#1750 ,

I can confirm that there is an issue with user models not named User or user due to hardcoded relations in userIdentity and userCredentials.

This falls under supporting extended user models. I have already submitted a PR fixing this here: https://github.com/strongloop/loopback-component-passport/pull/93

**Comment by jonathan-casarrubias** [comment]
Yes I can confirm I have same issue here, I always extend the user model and the component uses the base user model instead the one Im sending as userModel

**Comment by mbouclas** [comment]
Concerning #106 which was closed as duplicate. Anyone has any idea (or example) on how to auth (FB or Google) using a non standard model? By that i mean, a model that has extra required fields to the standard ones. For example, my model has a isActive field, and auth fails every time because of validation error, as the isActive does not exist in the FB response. In my case what i get after every auth attempt is this (or something in this lines)

```
422 ValidationError: The `Owner` instance is not valid. Details: `firstName` can't be blank (value: undefined); `lastName` can't be blank (value: undefined); `isActive` can't be blank (value: undefined).
```

**Comment by jonathan-casarrubias** [comment]
I think you simply should use an operation hook, Im doing something similar of what you try to achieve, I add a permalink depending the username facebook/twitter sent

````javascript
    Account.observe('before save', function(ctx, next) {
        if (!ctx.isNewInstance) { next(); }
        // I create a permalink and a relation model on before save hook
    });

    Account.observe('after save', function(ctx, next) {
        if (!ctx.isNewInstance) { next(); }
        // I send a welcome email message here
    });
````


The documentation states that it should run before validation

https://docs.strongloop.com/display/public/LB/Operation+hooks

Cheers,
@jonathan-casarrubias

**Comment by mbouclas** [comment]
@jonathan-casarrubias That brings me closer to what i'm after. Any idea if the ctx object holds the strategy data (FB return data)? I found an "instance" property but it does not hold the FB data. For example, i could use the first and last name returned by FB.

thanks

**Comment by jonathan-casarrubias** [comment]
The way I configured, I extended, User, UserIdentities and UserCredentials into Account, AccountIdentities and AccountCredentials, I explicitly added the relation as stated in the documentation, and I do have access to Account.identities() or ctx.instance.identitites() that includes the identities for that account so you may manipulate that data as your convenience.

In order for this to work, I needed to use the module I published that includes @clockworkgr's fix, which is the original topic of this thread, otherwise won't work, see above for more details.

Cheers,
@jonathan-casarrubias

**Comment by superkhau** [comment]
@loay Reassigning to you as you are the repo expert.

**Comment by loay** [comment]
Thanks. Having a look.

**Comment by ebarault** [comment]
@jonathan-casarrubias 
when you write : " I explicitly added the relation as stated in the documentation, and I do have access to Account.identities() or ctx.instance.identitites() that includes the identities for that account so you may manipulate that data as your convenience."

>> could you specify what relation exactly you added?
rigth now i cannot get identities in the operation hook before the custom user is created. empy array is always returned when invoking the relation.

thanks,

**Comment by btassone** [comment]
Hey guys any progress on this? I need to extend the user model in one of my projects and I wanted to use loopback 3.x. Sadly the fix up above is only slated for 2.x (the temporary fork is what I am talking about). Is there any timetable for when this will be fixed?

**Comment by jonathan-casarrubias** [comment]
I'm starting a project with loopback 3 and will need to use the component passport, I'll see how it goes and update here

**Comment by btassone** [comment]
@jonathan-casarrubias Any luck?

**Comment by jonathan-casarrubias** [comment]
@btassone we did start the project in LoopBack 3 and were able to use the fork version I published without problems.

Anyway, I had to take the decision of rolling back to version 2, not because of issues with this module, but because the version 3 alpha present many issues that the version 2 does not present, many of these were issues with HasMany relationships.

So, again we were able to use it in version 3, but we rolled back because other issues.

Hope this info is helpful for you

**Comment by mamartins** [comment]
1 year after and we're still here :p

**Comment by btassone** [comment]
You'd think a critical piece like this would get a bit more attention.

**Comment by loay** [comment]
@mamartins @btassone 
I will have a look today at the PR.

**Comment by btassone** [comment]
@loay My hero :D

**Comment by clockworkgr** [comment]
see https://github.com/strongloop/loopback-component-passport/pull/168 

needs rebasing but I'm currently in the middle of relocating for work so won't have time for a couple more weeks

**Comment by loay** [comment]
yeah I was about to wonder why did you close the main PR then I saw a replacement. I will try to rebase myself and see where it goes.

**Comment by loay** [comment]
here is an update: 
The pr needs units tests which is a must before merging to detect any regression. The rebasing will take some time too. The code itself looks fine to me logically, so I will try to work on the whole thing as soon as I can. I will keep you guys updated.

**Comment by mamartins** [comment]
Good to know!

In the mean time I developed my own solution, not as generic as this one but suits my needs :)

**Comment by btassone** [comment]
Thanks so much guys! We've had a few projects where this has been a huge roadblock so it will be nice not to run into this issue anymore. Thanks for your hard work!

**Comment by wprater** [comment]
any update on this one?  or is there a work around to allow a custom User model?

**Comment by nickvanvynckt** [comment]
@wprater, use the "loopback-component-passport-c" package someone mentioned here before. This allows you to use custom User models!

**Comment by lowell-list** [comment]
My custom User model (named PortalUser) is now being auto-created by `loopback-component-passport` with these two steps:

1) pass custom PortalUser model to `passportConfigurator.setupModels()`:
```
  // code in server.js
  passportConfigurator.setupModels({
    userModel: app.models.PortalUser,
    userIdentityModel: app.models.UserIdentity,
    userCredentialModel: app.models.UserCredential
  });
```

2) set the "belongsTo" relations of UserIdentity and UserCredential to point to PortalUser instead of User:
```
  ...
  // code in user-credential.json and user-identity.json
  "relations": {
    "user": {
      "type": "belongsTo",
      "model": "PortalUser",
      "foreignKey": "portalUserId"
    }
  }
  ...
```

**Comment by jonathan-casarrubias** [comment]
lol almost 3 years and same problems? that is why I decided to stop using loopback.. good luck guys

**Comment by skanygin** [comment]
@jonathan-casarrubias what are you using instead of it?

**Comment by rajkaran** [comment]
@lowell-list
This doesn't work for me
`"foreignKey": "portalUserId"`

have to change it to `"foreignKey": "userId"` because of
```
userIdentityModel.findOrCreate(
	{where: {externalId: profile.id}},
	{
	  provider: provider,
	  externalId: profile.id,
	  authScheme: authScheme,
	  profile: profile,
	  credentials: credentials,
	  userId: user.id,
	  created: date,
	  modified: date,
	},
	function(err, identity) {
	  if (!err && user && autoLogin) {
		return (options.createAccessToken || createAccessToken)(
		  user,
		  function(err, token) {
			cb(err, user, identity, token);
		  }
		);
	  }
	  cb(err, user, identity);
	}
);
```

**Comment by lowell-list** [comment]
@rajkaran - you are correct; I reviewed my code and discovered that I later changed `"foreignKey"` to `"userId"` in both portal-user-credential.json and portal-user-identity.json because I encountered the same issue you did. I updated my previous post with `"userId"` to help the next person.

**Comment by dragonxsx** [comment]
> My custom User model (named PortalUser) is now being auto-created by `loopback-component-passport` with these steps:
> 
>     1. Create custom user, user identity, and user credential models that extend from the provided User, UserIdentity, and UserCredential models. In my case I named them PortalUser, PortalUserIdentity, and PortalUserCredential.
> 
>     2. Set the "belongsTo" relations of PortalUserIdentity and PortalUserCredential to point to PortalUser instead of User:
> 
> 
> ```
>   ...
>   // code in portal-user-credential.json and portal-user-identity.json
>   "relations": {
>     "user": {
>       "type": "belongsTo",
>       "model": "PortalUser",
>       "foreignKey": "userId"
>     }
>   }
>   ...
> ```
> 
>     1. pass all three custom models through to `passportConfigurator.setupModels()`:
> 
> 
> ```
>   // code in server.js
>   passportConfigurator.setupModels({
>     userModel: app.models.PortalUser,
>     userIdentityModel: app.models.PortalUserIdentity,
>     userCredentialModel: app.models.PortalUserCredential
>   });
> ```

By refer to your solution, I overrided the 'user' relations of UserIdentity and UserCredential to make my CustomUser works.

```
  const MyCustomUser = app.models.MyCustomUser;
  const UserIdentity = app.models.UserIdentity;
  const UserCredential = app.models.UserCredential;

  // Fix problem related to custom userModel
  UserIdentity.belongsTo(MyCustomUser, {as: 'user', foreignKey: 'userId'});
  UserCredential.belongsTo(MyCustomUser, {as: 'user', foreignKey: 'userId'});

  const passportConfigurator = new PassportConfigurator(app);
  passportConfigurator.init();
  passportConfigurator.setupModels({
    userModel: MyCustomUser
  });
```
